fri/src/verifier.rs

```diff
@@ -1,38 +1,177 @@
     input_mmcs: &[FC::InputMmcs],
+    input_dims: &[Vec<Dimensions>],
+    input_commits: &[<FC::InputMmcs as Mmcs<FC::Val>>::Commitment],
     proof: &FriProof<FC>,
     challenger: &mut FC::Challenger,
-) -> Result<(), ()> {
-    for com in &proof.commit_phase_commits {
-        challenger.observe(com.clone());
-    }
+) -> VerificationResult<FC, ()> {
+    let alpha: FC::Challenge = challenger.sample_ext_element();
+    let betas: Vec<FC::Challenge> = proof
+        .commit_phase_commits
+        .iter()
+        .map(|comm| {
+            challenger.observe(comm.clone());
+            challenger.sample_ext_element()
+        })
+        .collect();
 
     if proof.query_proofs.len() != config.num_queries() {
-        return Err(());
+        return Err(VerificationError::InvalidProofShape);
     }
 
-    let log_max_height = 0; // TODO
-    let query_indices = (0..config.num_queries()).map(|_| challenger.sample_bits(log_max_height));
+    let log_max_height = proof.commit_phase_commits.len() + config.log_blowup();
 
-    for (index, proof) in query_indices.zip_eq(&proof.query_proofs) {
-        verify_query(config, input_mmcs, index, proof)?;
+    for query_proof in &proof.query_proofs {
+        let index = challenger.sample_bits(log_max_height);
+
+        let reduced_openings = verify_input(
+            input_mmcs,
+            input_commits,
+            input_dims,
+            &query_proof.input_openings,
+            index,
+            alpha,
+            log_max_height,
+        )?;
+
+        let folded_eval = verify_query(
+            config,
+            &proof.commit_phase_commits,
+            index,
+            query_proof,
+            &betas,
+            reduced_openings,
+            log_max_height,
+        )?;
+
+        if folded_eval != proof.final_poly {
+            return Err(VerificationError::FinalPolyMismatch);
+        }
     }
 
     Ok(())
 }
 
+fn verify_input<FC: FriConfig>(
+    input_mmcs: &[FC::InputMmcs],
+    input_commits: &[<FC::InputMmcs as Mmcs<FC::Val>>::Commitment],
+    input_dims: &[Vec<Dimensions>],
+    input_openings: &[InputOpening<FC>],
+    index: usize,
+    alpha: FC::Challenge,
+    log_max_height: usize,
+) -> VerificationResult<FC, Vec<FC::Challenge>> {
+    let mut openings_by_log_height: Vec<Vec<FC::Val>> = vec![vec![]; log_max_height + 1];
+    for (mmcs, commit, dims, opening) in
+        izip!(input_mmcs, input_commits, input_dims, input_openings)
+    {
+        mmcs.verify_batch(
+            commit,
+            dims,
+            index,
+            &opening.opened_values,
+            &opening.opening_proof,
+        )
+        .map_err(VerificationError::InputMmcsError)?;
+        for (mat_dims, mat_opening) in izip!(dims, &opening.opened_values) {
+            let log_height = log2_strict_usize(mat_dims.height);
+            openings_by_log_height[log_height].extend_from_slice(mat_opening);
+        }
+    }
+    let reduced_openings = openings_by_log_height
```

Comment:
BTW now that this is clean it should be easy to Horner eval the reduced FRI input instead. I'll do that in another PR that merges the reduction with the quotient as we talked about (`Mmcs::reduce_rows_with_ext`)